pyam_analysis/core.py

```diff
@@ -65,71 +65,94 @@ class IamDataFrame(object):
-            self.data = data.data
-            self._meta = data._meta
-            self.cat_color = data.cat_color
-            self.col_count = data.col_count
-
-        # read data from source
+        # import data from pandas.DataFrame or read from source
+        if isinstance(data, pd.DataFrame):
+            self.data = format_data(data)
+        elif has_ix and isinstance(data, ixmp.TimeSeries):
+            self.data = read_ix(data, regions=regions, variables=variables,
+                                units=units, years=years)
         else:
-            if has_ix and isinstance(data, ixmp.TimeSeries):
-                self.data = read_ix(data, regions)
-            else:
-                self.data = read_data(data, regions, **kwargs)
+            self.data = read_data(data, regions, **kwargs)
 
-            # define a dataframe for categorization and other meta-data
-            self._meta = return_index(self.data, mod_scen,
-                                      drop_duplicates=True)
-            self.reset_category(True)
+        # define a dataframe for categorization and other meta-data
+        self._meta = return_index(self.data, mod_scen,
+                                  drop_duplicates=True)
+        self.reset_category(True)
 
-            # define a dictionary for category-color mapping
-            self.cat_color = {'uncategorized': 'white', 'exclude': 'black'}
-            self.col_count = 0
+        # define a dictionary for category-color mapping
+        self.cat_color = {'uncategorized': 'white', 'exclude': 'black'}
+        self.col_count = 0
 
-    def append(self, other, regions=None):
-        """Read timeseries data and append to IamDataFrame
+    def append(self, other, inplace=False, regions=None, variables=None,
+               units=None, years=None, **kwargs):
+        """Import or read timeseries data and append to IamDataFrame
 
         Parameters
         ----------
-        data: ixmp.TimeSeries, ixmp.Scenario or data file
-            this option requires the 'ixmp' package as a dependency
-            or a file with IAMC-style data
-        regions: list
-            list of regions to be imported
+        other: pyam-analysis.IamDataFrame, ixmp.TimeSeries, ixmp.Scenario,
+        pandas.DataFrame or data file
+            an IamDataFrame, TimeSeries or Scenario (requires `ixmp`),
+            or pandas.DataFrame or data file with IAMC-format data columns
+        inplace : bool, default False
+            if True, do operation inplace and return None
+        regions : list of strings
+            filter by regions
+        variables : list of strings
+            filter by variables (only with ixmp objects)
+        units : list of strings
+            filter by units (only with ixmp objects)
+        years : list of integers
+            filter by years (only with ixmp objects)
         """
-        new = IamDataFrame(data=self)
+        new = copy.deepcopy(self)
```

Comment:
you should be able to say

```
new = copy.deepcopy(self) if not inplace else self
```

and ignore the inplace logic later on

Reply:
and we should probably develop a consistent style here. In other places I've used `ret` (short for `return`).